staging/src/k8s.io/apiextensions-apiserver/third_party/forked/celopenapi/model/schemas.go

```diff
@@ -151,11 +213,108 @@ func WithTypeAndObjectMeta(s *schema.Structural) *schema.Structural {
+// minimum size might be higher than what estimateMinSizeJSON returns.
+func estimateMinSizeJSON(s *schema.Structural) int64 {
+	if s == nil {
+		// minimum valid JSON token has length 1 (single-digit number like `0`)
+		return 1
+	}
+	switch s.Type {
+	case "boolean":
+		// true
+		return 4
+	case "number", "integer":
+		// 0
+		return 1
+	case "string":
+		if s.ValueValidation != nil {
+			switch s.ValueValidation.Format {
+			case "duration":
+				return minDurationSizeJSON
+			case "date":
+				return dateSizeJSON
+			case "date-time":
+				return minDatetimeSizeJSON
+			}
+		}
+		// ""
+		return 2
+	case "array":
+		// []
+		return 2
+	case "object":
+		// {}
+		objSize := int64(2)
+		// exclude optional fields since the request can omit them
+		if s.ValueValidation != nil {
+			for _, propName := range s.ValueValidation.Required {
+				if prop, ok := s.Properties[propName]; ok {
+					if prop.Default.Object != nil {
+						// exclude fields with a default, those are filled in server-side
+						continue
+					}
+					// add 4, 2 for quotations around the property name, 1 for the colon, and 1 for a comma
+					objSize += int64(len(propName)) + estimateMinSizeJSON(&prop) + 4
+				}
+			}
+		}
+		return objSize
+	}
+	if s.XIntOrString {
+		// 0
+		return 1
+	}
+	// this code should be unreachable, so return the safest possible value considering this can be used as
+	// a divisor
+	return 1
+}
+
+// estimateMaxArrayItemsPerRequest estimates the maximum number of array items with
+// the provided schema that can fit into a single request.
+func estimateMaxArrayItemsPerRequest(itemSchema *schema.Structural) int64 {
+	// subtract 2 to account for [ and ]
+	return (maxRequestSizeBytes - 2) / (estimateMinSizeJSON(itemSchema) + 1)
+}
+
+// estimateMaxStringLengthPerRequest estimates the maximum string length (in characters)
+// of a string compatible with the format requirements in the provided schema.
+// must only be called on schemas of type "string" or x-kubernetes-int-or-string: true
+func estimateMaxStringLengthPerRequest(s *schema.Structural) int64 {
+	if s.ValueValidation == nil || s.XIntOrString {
+		// subtract 2 to account for ""
+		return (maxRequestSizeBytes - 2)
+	}
+	switch s.ValueValidation.Format {
+	case "duration":
+		return maxDurationSizeJSON
+	case "date":
+		return dateSizeJSON
+	case "date-time":
+		return maxDatetimeSizeJSON
+	default:
+		// subtract 2 to account for ""
+		return (maxRequestSizeBytes - 2)
+	}
+}
+
+// estimateMaxAdditionalPropertiesPerRequest estimates the maximum number of additional properties
+// with the provided schema that can fit into a single request.
+func estimateMaxAdditionalPropertiesPerRequest(additionalPropertiesSchema *schema.Structural) int64 {
+	// 2 bytes for key + "" + colon + comma + smallest possible value, realistically the actual keys
+	// will all vary in length
+	// TODO(DangerOnTheRanger): is there a way to calculate how many bytes unique keys will need?
```

Comment:
Do you need a TODO for this?

Reply:
Probably not in retrospect, the 2 byte allowance should be fine. I'll remove the TODO and mark this as resolved once I do.